Enable debian packaging with https://packager.io

crohr opened the pull request:
This pull request adds support for packaging OpenProject for Debian 7 using Packager.io, and automatically setting up OpenProject with a wizard. It is not really meant to be merged as is, but to track the progress of the packaging process on the dev branch, and allow people to make comments.

I will extract a few separate pull requests from this branch as soon as I get back from vacation, to make the review and merging process easier.

see: https://www.openproject.org/work_packages/14571

crohr committed:
Allow to load ERB variables from config file.

crohr committed:
Generate fake secret token when precompiling.

crohr committed:
Allow to configure settings with rake task

crohr committed:
Add files required for packaging OpenProject with Packager.io

Conflicts:
	Gemfile.lock

crohr committed:
Use ADMIN_EMAIL to set default admin email

crohr committed:
Move unicorn definition to main Gemfile

crohr committed:
Add check script

crohr committed:
Save config

crohr committed:
Set serve_static_assets = true in case apache is not configured to serve assets.

crohr committed:
Add unicorn config.

crohr committed:
Use multi-buildpack to get nodejs and bower compilation.

crohr committed:
Launch bower install after npm install

crohr committed:
Add bower as proper dependency

crohr committed:
Set default web concurrency to 2

crohr committed:
Exit if sendmail can't be found

TeatroIO commented:
I've prepared a stage. <a href='http://crohr-dev.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Click to open.</a>

tessi commented on package.json:
```diff
@@ -28,7 +28,11 @@
   "scripts": {
-    "test": "./node_modules/karma/bin/karma start --single-run --browsers Firefox,PhantomJS"
+    "test": "./node_modules/karma/bin/karma start --single-run --browsers Firefox,PhantomJS",
+    "postinstall": "./node_modules/bower/bin/bower install"
```
@myabc May you have a look at this? I'm not confident enough to merge this part.

tessi commented on packaging/conf/configuration.yml:
```diff
@@ -0,0 +1,11 @@
+  smtp_authentication: :login
+  smtp_user_name: <%= ENV['SMTP_USERNAME'] %>
+  smtp_password: <%= ENV['SMTP_PASSWORD'] %>
+  attachments_storage_path: <%= ENV.fetch('ATTACHMENTS_STORAGE_PATH') { "/var/db/openproject/files" } %>
```
You can configure OpenProject just with environment variables. (without doing the erb-parsing). See: https://github.com/opf/openproject/blob/c9359dc31f224c3f3dffd0e55739f38ff497a9c7/doc/CONFIGURATION.md#environment-variables

tessi commented on packaging/debian/postinst:
```diff
@@ -0,0 +1,74 @@
+    ${CLI} config:set SECRET_TOKEN="$secret_token"
+
+    # migrate
+    ADMIN_EMAIL=$(${CLI} config:get ADMIN_EMAIL || echo "admin@example.com")
```
`|| echo "admin@example.com"` overwrites the default in seeds.rb - this is not DRY :) (meaning: if we want to change the default in seeds.rb for some reason, we have to remember to change this file too)

Besides that, you don't export the `ADMIN_EMAIL` variable so that the `rake db:seed` cannot see this variable.

tessi commented on packaging/debian/postinst:
```diff
@@ -0,0 +1,74 @@
+    ${CLI} config:set WEB_CONCURRENCY=${web_concurrency}
+
+    # set SECRET_TOKEN env variable
+    secret_token=$(${CLI} config:get SECRET_TOKEN || ( < /dev/urandom tr -dc a-z0-9 | head -c128; echo ))
```
I'd rather get the secret from openproject itself. We have a rake task for that

``` bash
$  bundle exec rake -s secret
require 'rails/all'... 0.360s
Bundler.require... 1.290s
afe678ba430b8b696273787aa1b30813531f581eba7e9b3a31f6297e99f245eb98f6b2c2d7be6b2419b3a0532ada7d52f81442639ac61c40f2e5a6fa7ad66444
```

I've just seen that the output of that task is a little verbose. We'll move the first two lines of output to `stderr` in a follow-up PR soon. (for now you can use `bundle exec rake -s secret | tail -1`)

tessi commented on packaging/scripts/backup:
```diff
@@ -0,0 +1,77 @@
+
+TARGET=${TARGET:=$ORIGINAL_PWD}
+TARGET=${TARGET:=$(pwd)}
+TARGET="/var/db/openproject/backup"
```
do you set `TARGET` three times here? I guess one time would be enough :)

mfrister commented on packaging/scripts/backup:
```diff
@@ -0,0 +1,77 @@
+mysqldump --defaults-file=${tmpfile} --single-transaction "${database}" | gzip > "$dst"
+echo " done" >&2
+echo "$dst"
+rm -f ${tmpfile}
```
This might be a bit picky: To properly clean up even if a command between creating and removing the temporary file fails, we should use [trap](http://www.linuxjournal.com/content/use-bash-trap-statement-cleanup-temporary-files).

Alternatively, [process substitution](http://www.tldp.org/LDP/abs/html/process-sub.html) could make this work without a temporary file:

``` bash
config=$(ruby -r ...)
mysqldump --defaults-file=<(echo $config) ...
```

tessi commented on packaging/scripts/backup:
```diff
@@ -0,0 +1,77 @@
+  fi
+else
+  echo "* No attachments folder. Ignoring." >&2
+fi
```
You could make this script more DRY by creating a function for the backup of a folder. They look _very_ similar to mee :)

mfrister commented on packaging/scripts/check:
```diff
@@ -0,0 +1,67 @@
+    log_ko "MySQL connection is NOT working"
+  fi
+
+  rm -f $tmpfile
```
Same as [here](https://github.com/opf/openproject/pull/1703/files#r15747390).

crohr committed:
Use correct group and user

crohr committed:
Use APP_GROUP instead of hardcoded openproject group

crohr committed:
Fix order of variables in postinstall

crohr committed:
Fix check script to use APP_USER

crohr commented:
Thanks to everyone for their reviews. I'm currently focusing on fixing the remaining blocking issues, and will get back to updating the PR based on your comments once the package is in a fully working state.

crohr committed:
Allow to overwrite rails_cache_store, session_store via environment variable

crohr committed:
Force ActiveRecord connection establishment with config/database.yml

This is necessary so that SSL settings are correctly used. By default,
Rails does not handle query parameters in the DATABASE_URL, while the
config/database.yml generated by buildpacks does.

crohr committed:
Simplify database check script

crohr committed:
Use APP_NAME everywhere

crohr committed:
Use APP_NAME everywhere, and handle ssl connection in backup script.

crohr committed:
Modify create-svn-repositories so that APP_NAME can be overwritten more
easily.

crohr committed:
Use APP_NAME everywhere in postinstall

crohr committed:
Can now switch package names and everything will work

crohr committed:
Fix path in clear old sessions cron

crohr committed:
Improve postinstall

kgalli commented:
This PR was merged into release/3.0 and will be merged into dev soon. See [PR 1857](https://github.com/opf/openproject/pull/1857). Thank you for your commitment.
